docs/source/building/variants.rst

```diff
@@ -0,0 +1,911 @@
+Build Variants
+==============
+
+
+The nature of binary compatibility (or lack thereof) means that unfortunately,
+we need to build binary packages (and any package containing binaries) with
+potentially several variants to support different usage environments. For
+example, using Numpy's C API means that a package must be used with the same
+version of Numpy as runtime that was used at build time.
+
+There has been limited support for this for a long time: Python in both build
+and run requirements resulted in a package that had python pinned to the version
+of Python used at build time (with a corresponding addition to the filename like
+py27). Similar support existed for numpy with the addition of an ``x.x`` pin in
+the recipe after `Conda-build PR
+573 <https://github.com/conda/conda-build/pull/573>`_ was merged. However, there
+has not been general support until conda-build version 3.0, though there have
+been many proposals (`Conda-build issue
+1142 <https://github.com/conda/conda-build/issues/1142>`_).
+
+As of conda-build 3.0, a new configuration scheme has been added, dubbed
+"variants." Conceptually, this decouples pinning values from recipes, replacing
+them with Jinja2 template variables. It adds support for the notion of
+"compatible" pinnings, though that concept is currently still under heavy
+development, to be integrated with ABI compatibility databases, such as `ABI
+Laboratory <https://abi-laboratory.pro/>`_.
+
+Variant input is ultimately a dictionary. They're mostly very flat. Keys are
+made directly available in Jinja2 templates. As a result, keys in the dictionary
+(and in files read into dictionaries) must be valid jinja2 variable names (no ``-``
+characters allowed). This example builds python 2.7 and 3.5 packages in one
+build command:
+
+conda_build_config.yaml like:
+
+.. code-block:: yaml
+
+   python:
+       - 2.7
+       - 3.5
+
+
+meta.yaml contents like:
+
+.. code-block:: yaml
+
+   package:
+       name: compiled-code
+       version: 1.0
+
+   requirements:
+       build:
+           - python {{ python }}
+       run:
+           - python {{ python }}
+
+
```

Comment:
How would you use variants to express the different combinations of compilers, processor-specific optimizations, and critical host/vendor packages (e.g. the mpi and blas/lapack libraries that are paired with the compiler toolchain?